packages/blocks/plus_minus/src/index.js

```diff
@@ -0,0 +1,1056 @@
+        !this.getInput('ELSE')
+    ) {
+      this.appendStatementInput('ELSE')
+          .appendField(Blockly.Msg['CONTROLS_IF_MSG_ELSE']);
+    }
+    this.updateShape_(targetCount);
+  },
+
+  /**
+   * Adds else-if and do inputs to the block until the block matches the
+   * target else-if count.
+   * @param {number} targetCount The target number of else-if inputs.
+   * @this Blockly.Block
+   * @private
+   */
+  updateShape_: function(targetCount) {
+    while (this.elseIfCount_ < targetCount) {
+      this.addPart_();
+    }
+    while (this.elseIfCount_ > targetCount) {
+      this.removePart_();
+    }
+  },
+
+  /**
+   * Callback for the plus field. Adds an else-if input to the block.
+   */
+  plus: function() {
+    this.addPart_();
+  },
+
+  /**
+   * Callback for the minus field. Triggers "removing" the input at the specific
+   * index.
+   * @see removeInput_
+   * @param {number} index The index of the else-if input to "remove".
+   * @this Blockly.Block
+   */
+  minus: function(index) {
+    if (this.elseIfCount_ == 0) {
+      return;
+    }
+    this.removePart_(index);
+  },
+
+  // To properly keep track of indices we have to increment before/after adding
+  // the inputs, and decrement the opposite.
+  // Because we want our first elseif to be IF1 (not IF0) we increment first.
+
+  /**
+   * Adds an else-if and a do input to the bottom of the block.
+   * @this Blockly.Block
+   * @private
+   */
+  addPart_: function() {
+    this.elseIfCount_++;
+    this.appendValueInput('IF' + this.elseIfCount_)
+        .setCheck('Boolean')
+        .appendField(Blockly.Msg['CONTROLS_IF_MSG_ELSEIF'])
+        .appendField(
+            new FieldMinus(this.elseIfCount_), 'MINUS' + this.elseIfCount_);
+    this.appendStatementInput('DO' + this.elseIfCount_)
+        .appendField(Blockly.Msg['CONTROLS_IF_MSG_THEN']);
+
+    // Handle if-elseif-else block.
+    if (this.getInput('ELSE')) {
+      this.moveInputBefore('ELSE', /* put at end */ null);
+    }
+  },
+
+  /**
+   * Appears to remove the input at the given index. Actually shifts attached
+   * blocks and then removes the input at the bottom of the block. This is to
+   * keep input names accurate.
+   * @param {number} index The index of the input to "remove".
+   * @this Blockly.Block
+   * @private
+   */
+  removePart_: function(index) {
+    // The strategy for removing a part at an index is to:
+    //  - Kick any blocks connected to the relevant inputs.
+    //  - Move all connect blocks from the other inputs up.
+    //  - Remove the last input.
+    // This makes sure all of our indices are correct.
+    if (index !== undefined) {
+      let foundInput = false;
+      const name = 'IF' + index;
+      for (let i = 0, input; (input = this.inputList[i]); i++) {
+        if (input.name == name) {
+          foundInput = true;
+          let connection = input.connection;
+          // TODO: Is there any reason we need to throw errors when the
```

Comment:
Is this TODO because that's the behaviour on the existing mutators in core Blockly?

Reply:
It is the existing behavior in core blockly, but I don't think any mutators use this. I believe they just let the blocks get kicked automatically when the input gets removed. But because we're faking which input gets removed, we have to kick them manually.